nebula2/gclient/net/__init__.py

```diff
@@ -33,34 +34,47 @@
 
 from nebula2.data.ResultSet import ResultSet
 
-logging.basicConfig(level=logging.INFO, format='[%(asctime)s]:%(message)s')
+logging.basicConfig(level=logging.INFO,
+                    format='[%(asctime)s] [%(filename)s:%(lineno)s]:%(message)s')
 
 
 class Session(object):
-    def __init__(self, connection, session_id, pool, retry_connect=True):
+    def __init__(self, address, session_id, pool, retry_connect=True):
         self.session_id = session_id
-        self._connection = connection
+        self._address = address
+        self._connection = None
         self._timezone = 0
         self._pool = pool
         self._retry_connect = retry_connect
+        self._lock = threading.RLock()
+        self._do_execute = False
+        self._do_safe_execute = False
 
     def execute(self, stmt):
         """
         execute statement
         :param stmt: the ngql
         :return: ResultSet
         """
-        if self._connection is None:
-            raise RuntimeError('The session has released')
+        with self._lock:
+            if self._do_safe_execute:
+                raise IOErrorException(
+                    IOErrorException.E_UNKNOWN,
+                    'The session is already called safe_execute, '
+                    'You can only use execute or safe_execute in the session')
```

Comment:
> You can only use execute or safe_execute in the session.
I don't understand this sentence.  Isn't it running the `execute` function now?

Reply:
A session can only be used by `execute` or `safe_execute`, not both at the same time, to avoid the waste of connection.